File: java-workshop/part-2-light-mocks/src/main/java/service/OrderService.java
package service;

import domain.Order;
import domain.OrderStatus;
import domain.PaymentResult;
import domain.PaymentStatus;
import domain.ShippingConfirmation;
import domain.StockReservation;
import repository.OrderRepository;

import java.util.List;

public class OrderService {
    private final OrderRepository orderRepository;
    private final StockService stockService;
    private final PricingService pricingService;
    private final PaymentService paymentService;
    private final ShippingService shippingService;
    private final DiscountCodeService discountCodeService;

    public OrderService(
        OrderRepository orderRepository,
        StockService stockService,
        PricingService pricingService,
        PaymentService paymentService,
        ShippingService shippingService,
        DiscountCodeService discountCodeService
    ) {
        this.orderRepository = orderRepository;
        this.stockService = stockService;
        this.pricingService = pricingService;
        this.paymentService = paymentService;
        this.shippingService = shippingService;
        this.discountCodeService = discountCodeService;
    }

    public OrderResult placeOrder(Order order) {
        if (!isValid(order)) {
            return new OrderResult.Failure(order.reject(), "Invalid order");
        }

        if (order.discountCode() != null
            && !discountCodeService.checkDiscountCode(order.customerId(), order.discountCode())) {
            return new OrderResult.Failure(order.reject(), "Discount code already used");
        }

        List<StockReservation> reservations = stockService.reserveStock(order.items());
        if (reservations.stream().anyMatch(reservation -> !reservation.reserved())) {
            return new OrderResult.Failure(order.reject(), "Insufficient stock");
        }

        double totalPrice = order.discountCode() != null
            ? pricingService.calculateTotal(order.items(), order.discountCode())
            : pricingService.calculateTotal(order.items());
        PaymentResult paymentResult = paymentService.processPayment(
            order.id(),
            order.customerId(),
            totalPrice
        );

        if (paymentResult.status() != PaymentStatus.SUCCESS) {
            stockService.releaseStock(reservations);
            return new OrderResult.Failure(order.reject(), "Payment failed");
        }

        ShippingConfirmation shippingConfirmation = shippingService.createShipment(order);
        Order confirmedOrder = order.confirm(totalPrice);
        orderRepository.save(confirmedOrder);

        if (order.discountCode() != null) {
            discountCodeService.markAsUsed(order.customerId(), order.discountCode());
        }

        return new OrderResult.Success(confirmedOrder, shippingConfirmation);
    }

    private boolean isValid(Order order) {
        return order.status() == OrderStatus.PENDING
            && order.items() != null
            && !order.items().isEmpty();
    }
}

File: typescript-workshop/part-1-module-mocks/src/service/OrderService.ts
import { DiscountCode } from '../domain/DiscountCode';
import { Order } from '../domain/Order';
import { OrderStatus } from '../domain/OrderStatus';
import { PaymentStatus } from '../domain/PaymentStatus';
import { OrderRepository } from '../repository/OrderRepository';
import { calculateTotal, calculateTotalWithDiscount } from './PricingService';
import { OrderResult } from './OrderResult';
import { PaymentService } from './PaymentService';
import { ShippingService } from './ShippingService';
import { StockService } from './StockService';

export class OrderService {
  placeOrder(order: Order, discountCode?: DiscountCode): OrderResult {
    const stockService = new StockService();
    const paymentService = new PaymentService();
    const shippingService = new ShippingService();
    const orderRepository = new OrderRepository();

    if (!this.isValid(order)) {
      return OrderResult.failure(order.reject(), 'Invalid order');
    }

    const reservations = stockService.reserveStock(order.items);
    if (reservations.some((r) => !r.reserved)) {
      return OrderResult.failure(order.reject(), 'Insufficient stock');
    }

    const totalPrice = discountCode !== undefined
      ? calculateTotalWithDiscount(order.items, discountCode)
      : calculateTotal(order.items);

    const paymentResult = paymentService.processPayment(order.id, order.customerId, totalPrice);
    if (paymentResult.status !== PaymentStatus.Success) {
      stockService.releaseStock(reservations);
      return OrderResult.failure(order.reject(), 'Payment failed');
    }

    const shippingConfirmation = shippingService.createShipment(order);
    const confirmedOrder = order.confirm(
      totalPrice,
      paymentResult.transactionId,
      shippingConfirmation.trackingNumber,
    );

    orderRepository.save(confirmedOrder);
    return OrderResult.success(confirmedOrder, shippingConfirmation);
  }

  isValid(order: Order): boolean {
    return order.status === OrderStatus.Pending && order.items.length > 0;
  }
}
